### ISSUE TITLE
Expose target-cpu=native as a profile setting

### ISSUE THREAD

**Issue by bluss** [rust-lang/cargo#2535]
For a lot of projects you want that the compiler uses every cpu feature available. Not just numerics, but loops that copy or search in general can profit from using wider registers or vector instructions.

target-cpu=native makes plain old Rust code much more powerful. We don't need to write simd intrinsics, simd types, we can write one version of the code that works everywhere but the compiler can compile it well. It's not perfect, not always a replacement for hand implemented vectorization, but it's pretty good.

As of this writing, [matrixmultiply](https://github.com/bluss/matrixmultiply) is in plain rust. It should compile with any settings, but using `-C target-cpu=native` is a big benefit (up to 50% increase in throughput if you have avx). It would therefore be great if projects using it had an easy way to just flip the switch that they want the compiler to use all available cpu features.

**Comment by bluss** [comment]
Re https://github.com/rust-lang/cargo/issues/2112#issuecomment-192362539 this falls between the two scenarios in the comment. Neither testing nor required, just a very good thing to use when it's available.

**Comment by alexcrichton** [comment]
Yeah I don't think we've quite figured out a great way to deal with these sorts of options today. There's `RUSTFLAGS` which can suffice in the meantime, but it's not a great interface unfortunately.

To me this is essentially the same question of exposing `target-feature` in Cargo to crates somehow. Ideally, though, the entire DAG is compiled with the same options (including the standard library) which is the hard part.

**Comment by retep998** [comment]
I'd really like if I could set this to be the default for _all_ my builds locally since I typically don't care about distributing binaries.

**Comment by bluss** [comment]
@retep998 That seems to be possible using build.rustflags in the user-wide cargo config.

**Comment by HadrienG2** [comment]
As discussed in #16051, there is another reason why Cargo would benefit from first-class handling of `target-cpu=native`, which is that it would allow Cargo to correctly handle build cache invalidation when build directories are migrated across machines due to CI caching, simple rsync/tarball/scp data transfers that include the `target/` directory, etc.

Interestingly, this correctness concern is specific to `target-cpu=native` and not shared with other related flags such as `target-cpu=znver2` or `target-feature=+avx512f`. For those, the current Cargo cache invalidation scheme of rebuilding everything whenever rustflags change is sufficient. That being said, from the perspective of performance-focused users, it would still make more sense to expose all of these settings in the same place, i.e. make all of them handled by cargo or none of them as is currently the case.
